src/ripple/overlay/MessageScheduler.h

```diff
@@ -0,0 +1,451 @@
+{
+public:
+    enum class FailureCode {
+        // The request timed out.
+        TIMEOUT,
+        // The peer disconnected.
+        DISCONNECT,
+        // The message scheduler is shutting down.
+        SHUTDOWN,
+    };
+
+    using RequestId = uint32_t;
+
+    class Offer;
+
+    class Sender;
+
+    // A receiver has methods for success and failure callbacks.
+    // We package them together in the same object because they may share data,
+    // and we want to make it easy to control the lifetime of that data.
+    class Receiver;
+
+    // The `request_cookie` field on a `TMLedgerData` message represents
+    // a peer ID, which is a low number starting at 1 and believed to never
+    // exceed 300. By using a non-overlapping set of numbers for our request
+    // IDs, `PeerImp` can dispatch incoming responses `TMLedgerData` messages to
+    // the correct receiver. Similarly for the `seq` field on
+    // `TMGetObjectByHash` messages.
+    static constexpr RequestId MINIMUM_REQUEST_ID = 1 << 9;
+
+private:
+    using Clock = std::chrono::steady_clock;
+    using Timer = boost::asio::basic_waitable_timer<Clock>;
+    using Time = std::chrono::time_point<Clock>;
+
+    // An in-flight request.
+    struct Request
+    {
+        RequestId id;
+        Channel channel;
+        Receiver* receiver;
+        Timer timer;
+        Time sent;
+    };
+
+    boost::asio::io_service& io_service_;
+    beast::Journal journal_;
+
+    // Offers are negotiations between peers and senders.
+    // This mutex must be locked when handling either set.
+    std::mutex offersMutex_;
+    std::vector<Channel> channels_;
+    // TODO: Use a priority queue.
+    std::vector<Sender*> senders_;
+    bool stopped_ = false;
+
+    // Randomize the first ID to avoid collisions after a restart.
+    // C++23: We may be able to switch to std::randint.
+    // https://en.cppreference.com/w/cpp/experimental/randint
+    std::atomic<RequestId> nextId_ =
+        rand_int<RequestId>(MINIMUM_REQUEST_ID, 1 << 24);
+    std::mutex requestsMutex_;
+    // TODO: Might make sense to use a set instead.
+    hash_map<RequestId, std::unique_ptr<Request>> requests_;
+
+public:
+    MessageScheduler(
+        boost::asio::io_service& io_service,
+        beast::Journal const& journal)
+        : io_service_(io_service), journal_(journal)
+    {
+    }
+
+    /**
+     * If there are any waiting senders, offer these channels to them.
+     * Add any remaining open channels to the pool.
+     *
+     * TODO: Maybe this can be private.
+     */
+    void
+    connect(std::vector<Channel> channels);
+
+    void
+    connect(std::shared_ptr<Peer> peer)
+    {
+        JLOG(journal_.trace())
+            << "connect,id=" << peer->id()
+            << ",address=" << peer->getRemoteAddress().to_string();
+        Channel channel{peer->id(), peer};
+        // TODO: Let a peer choose its number of channels when it connects.
+        constexpr std::size_t CHANNELS_PER_PEER = 1;
```

Comment:
For now, the only user of `MessageScheduler` is my new `InboundLedger`. In my tests, trying more than one simultaneous request results in disconnects. We might instead need a more sophisticated governor that measures number of bytes transmitted instead of number of messages. I think it will be fine gating the migration of other components to using the `MessageScheduler` on a more flexible governor, but I'd rather not gate the PR as long as the only component using it is `InboundLedger`.

Reply:
Sounds reasonable. We do need to gate the migration of other components to using the `MessageScheduler`.